express-notes-api/server.ts

```diff
@@ -0,0 +1,154 @@
+      error.error = `${req.params.id} is not a valid id number.`;
+      res.status(400).json(error);
+      throw new Error(`${req.params.id} is not a valid id number.`);
+    }
+    if (!data.notes[id]) {
+      error.error = `Note with ID ${id} does not exist.`;
+      res.status(404).json(error);
+      throw new Error(`Note with ID ${id} does not exist.`);
+    }
+    res.status(200).json(data.notes[id]);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+// Creates a new note entry with a unique id and fills in the content given
+app.post('/api/notes', async (req, res) => {
+  try {
+    await readJSON();
+    if (!req.body.content) {
+      error.error = 'Must include content property.';
+      res.status(400).json(error);
+      throw new Error('Must include content property.');
+    }
+    const id = data.nextId++;
+    data.notes[id] = {
+      id,
+      content: req.body.content,
+    };
+    await writeJSON(res);
+    res.status(201).json(data.notes[id]);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+// Removes a note based on the id of the note
+app.delete('/api/notes/:id', async (req, res) => {
+  try {
+    await readJSON();
+    const id: number = Number(req.params.id);
+    if (isNaN(id) || id < 1 || !Number.isInteger(id)) {
+      error.error = `${id} is not a valid id number.`;
+      res.status(400).json(error);
+      throw new Error(`${id} is not a valid id number.`);
+    }
+    if (!data.notes[id]) {
+      error.error = `Note with ID ${id} does not exist.`;
+      res.status(404).json(error);
+      throw new Error(`Note with ID ${id} does not exist.`);
+    }
+    delete data.notes[id];
+    await writeJSON(res);
+    res.sendStatus(204);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+// Amends a pre-exisiting note with a new content field
+app.put('/api/notes/:id', async (req, res) => {
+  try {
+    await readJSON();
+    const id: number = Number(req.params.id);
+    if (isNaN(id) || id < 1 || !Number.isInteger(id) || !req.body.content) {
+      error.error = `${id} must be a valid number and content field must be included.`;
+      res.status(400).json(error);
+      throw new Error(
+        `${id} must be a valid number and content field must be included.`
+      );
+    }
+    if (!data.notes[id]) {
+      error.error = `Note with ID ${id} does not exist.`;
+      res.status(404).json(error);
+      throw new Error(`Note with ID ${id} does not exist.`);
+    }
+    data.notes[id] = {
+      id,
+      content: req.body.content,
+    };
+    await writeJSON(res);
+    res.status(200).json(data.notes[id]);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+app.listen(8080, () => {
+  console.log('Express server listening on port 8080.');
+});
+
+async function readJSON() {
+  try {
+    data = JSON.parse(await readFile('data.json', 'utf8'));
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+async function writeJSON(res: express.Response) {
```

Comment:
This should take `data` as a parameter. See above.

Reply:
I believe I fixed everything. I think I just had my original error handling set up backwards from how you suggested. Switched it around so the responses are in the catch blocks of the method handlers and the read/write functions will propagate the exception up to those handler functions.